File: src/gcs/bench.py
"""Benchmark: solve times, plan replay and drag frame rates.

    python -m gcs.bench

Solve times are for a *compiled* System (what dragging pays per frame) and for compile+solve
(one-shot).  Kill background CPU hogs before trusting numbers.
"""

from __future__ import annotations

import time
from collections.abc import Callable

import numpy as np

from gcs.decompose import PlanDrag, PlanSolver
from gcs.examples import EXAMPLES, perturb, truss, truss_floating, zigzag
from gcs.model import Sketch
from gcs.solve import METHODS, Drag, Method, System

MakeSketch = Callable[[], Sketch]


def bench_solve(make: MakeSketch, method: Method, reps: int = 20,
                sigma: float = 1.0) -> tuple[float, bool, float]:
    """(median compiled-solve ms, all ok, mean iterations)."""
    ts, ok, its = [], True, 0.0
    for i in range(reps):
        sk = make()
        perturb(sk, sigma, seed=i)
        s = System(sk)
        t0 = time.perf_counter()
        r = s.solve(method=method)
        ts.append(time.perf_counter() - t0)
        s.dispose()
        ok &= r.success
        its += r.iterations
    return float(np.median(ts)) * 1e3, ok, its / reps


def bench_compile(make: MakeSketch, reps: int = 20) -> float:
    sk = make()
    ts = []
    for _ in range(reps):
        t0 = time.perf_counter()
        System(sk).dispose()
        ts.append(time.perf_counter() - t0)
    return float(np.median(ts)) * 1e3


def bench_drag(sk: Sketch, frames: int = 20) -> tuple[float, bool]:
    p = sk.points[len(sk.points) // 2]
    d = Drag(sk, p, *p.xy)
    ts = []
    res = None
    for _ in range(frames):
        res = d.move(p.xy[0] + 1.0, p.xy[1] + 0.5)
        ts.append(res.time_s)
    d.end()
    return float(np.median(ts)) * 1e3, bool(res and res.success)


def bench_plan_drag(sk: Sketch, point_index: int, frames: int = 20,
                    plan: PlanSolver | None = None) -> tuple[float, float]:
    """(drag start ms, median frame ms) of the app's drag on `point_index` — on the sketch's
    cached plan when given one, as the app drags, else with a plan of its own."""
    p = sk.points[point_index]
    x, y = p.xy
    t0 = time.perf_counter()
    d = PlanDrag(sk, p, x, y, plan=plan)
    start = time.perf_counter() - t0
    ts = []
    for i in range(frames):
        t0 = time.perf_counter()
        d.move(x + 2.0 * np.cos(0.3 * i), y + 2.0 * np.sin(0.3 * i))
        ts.append(time.perf_counter() - t0)
    d.end()
    return start * 1e3, float(np.median(ts)) * 1e3


def main() -> None:
    cases: dict[str, MakeSketch] = dict(EXAMPLES)
    cases["truss_50"] = lambda: truss(bays=50)
    cases["truss_100"] = lambda: truss(bays=100)
    print("== solve (perturbed warm start): compiled-solve ms / iterations ==")
    print(f"{'sketch':<14}{'free':>5}{'res':>5} | " + " | ".join(f"{m:^16}" for m in METHODS)
          + " | compile")
    for name, make in cases.items():
        sk = make()
        big = len(sk.params) > 100
        cells = []
        for m in METHODS:
            ms, ok, it = bench_solve(make, m, reps=5 if big else 20)
            cells.append(f"{ms:7.2f} ms {'' if ok else 'BAD'} it={it:4.1f}")
        print(f"{name:<14}{len(sk.free_indices()):>5}{sk.n_residuals():>5} | " + " | ".join(cells)
              + f" | {bench_compile(make):5.2f} ms", flush=True)

    print("\n== decomposition plan (Stage 3): compile once, replay per solve ==")
    for name, make in cases.items():
        sk = make()
        t0 = time.perf_counter()
        ps = PlanSolver(sk)
        tc = time.perf_counter() - t0
        te = []
        for i in range(5):
            perturb(sk, 1.0, seed=i)
            t0 = time.perf_counter()
            ps.execute()
            te.append(time.perf_counter() - t0)
        r = ps.solve(fallback=False)
        print(f"{name:<14} compile {tc * 1e3:7.1f} ms | replay "
              f"{float(np.median(te)) * 1e3:7.2f} ms | {ps.plan.summary} | "
              f"{'exact' if r.success else 'needs fallback'}", flush=True)
        ps.dispose()

    print("\n== drag frame (pull + polish), dogleg ==")
    for bays in (30, 50, 100, 200):
        sk = truss(bays)
        ents = len(sk.points) + len(sk.lines)
        ms, ok = bench_drag(sk)
        ms2, ok2 = bench_drag(truss_floating(bays))
        print(f"truss({bays:3d}) {ents:5d} entities: fully constrained {ms:6.1f} ms "
              f"({1e3 / ms:4.0f} fps) | floating rigid {ms2:6.1f} ms ({1e3 / ms2:4.0f} fps) "
              f"{'ok' if ok and ok2 else 'BAD'}", flush=True)

    print("\n== drag of one figure among many (PlanDrag start + frame): the cost of the region ==")
    print("   own plan: the drag decomposes the figure | cached plan: as the app drags")
    for n, copies in ((32, 1), (32, 3), (32, 30), (128, 1), (2048, 1)):
        sk = zigzag(n, copies)
        start, frame = bench_plan_drag(sk, n // 2)   # a point of the first staircase
        ps = PlanSolver(sk, sticky=True)
        ps.solve()
        start2, frame2 = bench_plan_drag(sk, n // 2, plan=ps)
        ps.dispose()
        print(f"zigzag {n:4d} x {copies:2d} ({len(sk.points):5d} points): own plan start {start:7.2f} ms "
              f"frame {frame:6.3f} ms | cached plan start {start2:6.2f} ms frame {frame2:6.3f} ms",
              flush=True)


if __name__ == "__main__":
    main()

File: rust/gcs-core/src/bin/bench.rs
//! Benchmark: solve times, plan replay and drag frame rates — the native build.
//!
//!     cargo run --release -p gcs-core --bin bench
//!
//! Solve times are for a *compiled* `System` (what dragging pays per frame) and for
//! compile + solve (one-shot).  `npm run bench` measures the same things through wasm; this is
//! the native half, and the two are meant to be read side by side.
//!
//! Wall-clock medians over repetitions, and nothing else: the core has no dependencies and a
//! benchmark harness would be the first.  Kill background CPU hogs before trusting the numbers —
//! `uptime` first.

use std::time::Instant;

use gcs_core::decompose::{PlanDrag, PlanSolver};
use gcs_core::examples::{self, truss, truss_floating, zigzag};
use gcs_core::model::Sketch;
use gcs_core::newton::Method;
use gcs_core::solve::{Drag, SolveOpts};
use gcs_core::system::System;

const METHODS: [Method; 2] = [Method::DogLeg, Method::Lm];

fn median(mut v: Vec<f64>) -> f64 {
    if v.is_empty() {
        return f64::NAN;
    }
    v.sort_by(|a, b| a.partial_cmp(b).unwrap());
    let n = v.len();
    if n % 2 == 1 {
        v[n / 2]
    } else {
        0.5 * (v[n / 2 - 1] + v[n / 2])
    }
}

fn ms(t: Instant) -> f64 {
    t.elapsed().as_secs_f64() * 1e3
}

/// (median compiled-solve ms, all ok, mean iterations)
fn bench_solve(make: &dyn Fn() -> Sketch, method: Method, reps: u32) -> (f64, bool, f64) {
    let (mut ts, mut ok, mut its) = (Vec::new(), true, 0.0);
    for i in 0..reps {
        let mut sk = make();
        examples::jitter(&mut sk, 1.0, i + 1);
        let mut s = System::new(&sk);
        let t0 = Instant::now();
        let r = s.solve(&mut sk, SolveOpts { method, ..SolveOpts::default() });
        ts.push(ms(t0));
        ok &= r.success;
        its += r.iterations as f64;
    }
    (median(ts), ok, its / reps as f64)
}

/// (median compile ms, free columns, residual rows) — the dimensions come off the last system
/// it built, so nothing compiles a `System` twice to print two integers.
fn bench_compile(sk: &Sketch, reps: u32) -> (f64, usize, usize) {
    let mut ts = Vec::new();
    let (mut free, mut res) = (0, 0);
    for _ in 0..reps {
        let t0 = Instant::now();
        let s = System::new(sk);
        ts.push(ms(t0));
        (free, res) = (s.n_free, s.n_res);
    }
    (median(ts), free, res)
}

fn bench_drag(mut sk: Sketch, frames: u32) -> (f64, bool) {
    let p = sk.points.len() / 2;
    let (x, y) = sk.point_xy(p);
    let mut d = Drag::new(&mut sk, p, x, y, Method::DogLeg, 1.0, Vec::new(), 0.05);
    let mut ts = Vec::new();
    let mut ok = false;
    // The cursor has to keep moving.  Held at one target the drag reaches it on the first frame
    // and the rest time a solve with nothing to do — which is what printed 30,000 fps.
    for i in 0..frames {
        let a = 0.3 * i as f64;
        let t0 = Instant::now();
        let r = d.move_to(&mut sk, x + 1.0 + a.cos(), y + 0.5 + a.sin());
        ts.push(ms(t0));
        ok = r.success;
    }
    d.end(&mut sk);
    (median(ts), ok)
}

/// (drag start ms, median frame ms) of the app's drag on `point` — on the sketch's cached plan
/// when given one, as the app drags, else with a plan of its own.
fn bench_plan_drag(sk: &mut Sketch, point: usize, frames: u32, own: bool) -> (f64, f64) {
    let (x, y) = sk.point_xy(point);
    // The cached plan is the app's: decomposed and solved by the edit before the gesture, so
    // neither cost lands in `start`.  Its absence is exactly what the `own` column measures.
    let mut plan = (!own).then(|| {
        let mut p = PlanSolver::new(sk, true);
        p.solve(sk, 1e-9, true, Method::DogLeg);
        p
    });
    let t0 = Instant::now();
    let mut d = match plan.as_mut() {
        Some(p) => PlanDrag::on(sk, p, point, x, y, None, 0.05),
        None => PlanDrag::new(sk, point, x, y, None, 0.05),
    };
    let start = ms(t0);
    let cached = plan.as_ref().map(|p| &p.plan);
    let mut ts = Vec::new();
    for i in 0..frames {
        let a = 0.3 * i as f64;
        let t0 = Instant::now();
        d.move_to(sk, cached, x + 2.0 * a.cos(), y + 2.0 * a.sin());
        ts.push(ms(t0));
    }
    d.end();
    (start, median(ts))
}

type Case = (&'static str, Box<dyn Fn() -> Sketch>);

fn cases() -> Vec<Case> {
    let mut v: Vec<Case> = Vec::new();
    for name in examples::EXAMPLES {
        v.push((name, Box::new(move || examples::example(name).expect("example"))));
    }
    v.push(("truss_50", Box::new(|| truss(50, 20.0, 15.0, true))));
    v.push(("truss_100", Box::new(|| truss(100, 20.0, 15.0, true))));
    v
}

fn main() {
    let cases = cases();

    println!("== solve (jittered warm start): compiled-solve ms / iterations ==");
    print!("{:<16}{:>5}{:>5} |", "sketch", "free", "res");
    for m in METHODS {
        print!(" {:^24} |", m.as_str());
    }
    println!(" compile");
    for (name, make) in &cases {
        let sk = make();
        let reps = if sk.params.len() > 100 { 5 } else { 20 };
        let mut cells = Vec::new();
        for m in METHODS {
            let (t, ok, it) = bench_solve(make.as_ref(), m, reps);
            cells.push(format!("{t:8.2} ms {:3} it={it:4.1}", if ok { "" } else { "BAD" }));
        }
        let (tc, free, res) = bench_compile(&sk, reps);
        print!("{name:<16}{free:>5}{res:>5} |");
        for c in cells {
            print!(" {c} |");
        }
        println!(" {tc:5.2} ms");
    }

    println!("\n== decomposition plan: compile once, replay per solve ==");
    for (name, make) in &cases {
        let mut sk = make();
        let t0 = Instant::now();
        let mut ps = PlanSolver::new(&sk, false);
        let tc = ms(t0);
        let mut te = Vec::new();
        for i in 0..5 {
            examples::jitter(&mut sk, 1.0, i + 1);
            let t0 = Instant::now();
            gcs_core::decompose::execute(&mut ps.plan, &mut sk, None);
            te.push(ms(t0));
        }
        let r = ps.solve(&mut sk, 1e-9, false, Method::DogLeg);
        println!(
            "{name:<16} compile {tc:7.1} ms | replay {:7.2} ms | {} | {}",
            median(te),
            ps.plan.summary(),
            if r.success { "exact" } else { "needs fallback" }
        );
    }

    println!("\n== drag frame (pull + polish), dogleg ==");
    for bays in [30usize, 50, 100, 200] {
        let sk = truss(bays, 20.0, 15.0, true);
        let ents = sk.points.len() + sk.lines.len();
        let (t, ok) = bench_drag(sk, 20);
        let (t2, ok2) = bench_drag(truss_floating(bays), 20);
        println!(
            "truss({bays:3}) {ents:5} entities: fully constrained {t:6.1} ms ({:4.0} fps) | \
             floating rigid {t2:6.1} ms ({:4.0} fps) {}",
            1e3 / t,
            1e3 / t2,
            if ok && ok2 { "ok" } else { "BAD" }
        );
    }

    println!("\n== drag of one figure among many (PlanDrag start + frame): the cost of the \
              region ==");
    println!("   own plan: the drag decomposes the figure | cached plan: as the app drags");
    for (n, copies) in [(32usize, 1usize), (32, 3), (32, 30), (128, 1), (2048, 1)] {
        let mut sk = zigzag(n, copies);
        let npts = sk.points.len();
        let (start, frame) = bench_plan_drag(&mut sk, n / 2, 20, true);
        let mut sk = zigzag(n, copies); // from the same geometry as the row above
        let (start2, frame2) = bench_plan_drag(&mut sk, n / 2, 20, false);
        println!(
            "zigzag {n:4} x {copies:2} ({npts:5} points): own plan start {start:7.2} ms \
             frame {frame:6.3} ms | cached plan start {start2:6.2} ms frame {frame2:6.3} ms"
        );
    }
}
